### PR TITLE
[ZEN-4711] OpenApiDoc: Rendering Error on top-level OAuth Security Scheme

### PR DESCRIPTION
**Pull request by ghillairet**
This PR contains a fix in the Normalizer to retain SecuritySchemes when used in DocView.

Actual fix https://github.com/RepreZen/GenFlow/pull/46/files#diff-84180cae29f56fe5e5f34ff66c3d727eR199

### CODE REVIEW

File: modules/genflow-api/src/test/java/com/reprezen/genflow/api/normal/openapi/OpenApiNormalizerTest.xtend [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+		val tree = mapper.readTree(content)
+
+		new OpenApiNormalizer(
+			OPENAPI3_MODEL_VERSION,
+			Option.DOC_DEFAULT_OPTIONS
+		).of(tree).normalizeToKaizen(new URL("file://test.yaml")) as OpenApi3
+	}
+
+	@Test
+	def void testRetainSecuritySchemesInDocView() {
```
Comment (tedepstein) [inline modules/genflow-api/src/test/java/com/reprezen/genflow/api/normal/openapi/OpenApiNormalizerTest.xtend]: @ghillairet , nice! 
Testing is good.

**Comment by ghillairet** [comment]
@tedepstein Please look at the last commit. I implemented a solution that adds a ROOT ObjectType. Thanks to this we can access to `/security` array when we want to retain implicitly referenced objects (as done here https://github.com/RepreZen/GenFlow/pull/46/commits/fc7ca5649958ea0e10a4243d0c22f8b5b98a93ac#diff-84180cae29f56fe5e5f34ff66c3d727eR204).

File: modules/genflow-api/src/test/java/com/reprezen/genflow/api/normal/openapi/OpenApiNormalizerTest.xtend [hunk 2]
```diff
@@ -52,9 +54,57 @@ class OpenApiNormalizerTest {
 		'''
 		
 		val result = parse(content)
-		assertNotNull(result)
+
+		assertTrue(result.isValid)
 
 		assertNotNull(result.securitySchemes.get("oAuth2ClientCredentials"))
+		assertNull(result.securitySchemes.get("notUsedScheme"))
```
Comment (tedepstein) [inline modules/genflow-api/src/test/java/com/reprezen/genflow/api/normal/openapi/OpenApiNormalizerTest.xtend]: 👍

File: modules/genflow-api/src/test/java/com/reprezen/genflow/api/normal/openapi/OpenApiNormalizerTest.xtend [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+            info:
+              title: ''
+              version: ''
+            paths:
+              /entry:
+                get:
+                  security:
+                    - oAuth2ClientCredentials:
+                      - auth
+                  responses:
```
Comment (tedepstein) [inline modules/genflow-api/src/test/java/com/reprezen/genflow/api/normal/openapi/OpenApiNormalizerTest.xtend]: 👍

**Comment by tedepstein** [comment]
... @ghillairet , please disregard my previous comment about a failed build. The build is fine, and I will QA this now.

**Comment by ghillairet** [comment]
@tedepstein PR updated to include your suggestions.

**Comment by tedepstein** [comment]
Note: I want to run a final QA check on this to make sure it handles externally referenced Security Schemes correctly, or at least does not introduce any new errors related to this.

**Comment by tedepstein** [comment]
This failed QA. 

I tested with [build  #1.6.16.10050](http://build.reprezen.com:8111/viewLog.html?buildId=7883&buildTypeId=RepreZen_DevelopmentBuild&tab=artifacts
) from ModelSolv/RepreZen [PR #1857](https://github.com/ModelSolv/RepreZen/pull/1857).

On OpenAPI v2 and v3 documents, the live Documentation view shows a Rendering Error, and the Documentation GenTemplates fail without only a generic error message, nothing relevant to this particular issue, and no console output. (That may be a separate issue.) 

The live view Rendering Error puts an exception in the error log: 

```
Failed to generate content for live view: Documentation View: Failed to parse Swagger spec: C:\Users\tedep\RepreZen\workspace\WS.1.6.16.10050\Pet Store (v3)\models\rzn6339960675825541290json

com.reprezen.genflow.api.GenerationException: Failed to parse Swagger spec: C:\Users\tedep\RepreZen\workspace\WS.1.6.16.10050\Pet Store (v3)\models\rzn6339960675825541290json
	at com.reprezen.genflow.api.openapi3.OpenApi3Source.load(OpenApi3Source.java:54)
	at com.reprezen.genflow.api.openapi3.OpenApi3Source$OpenApi3Source_DocNormalizerOptions.load(OpenApi3Source.java:162)
	at com.reprezen.genflow.api.source.AbstractSource.load(AbstractSource.java:46)
	at com.reprezen.genflow.api.template.GenTemplate$StaticGenerator.generate(GenTemplate.java:462)
	at com.reprezen.genflow.api.template.GenTemplate$Generator.generate(GenTemplate.java:763)
	at com.reprezen.genflow.api.target.GenTarget.execute(GenTarget.java:244)
	at com.reprezen.genflow.api.target.GenTarget.execute(GenTarget.java:223)
	at com.modelsolv.reprezen.gentemplates.common.ui.newlive.LiveGeneratorWrapper.generate(LiveGeneratorWrapper.java:101)
	at com.modelsolv.reprezen.gentemplates.common.ui.view.BrowserViewPage$2.run(BrowserViewPage.java:245)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:56)
Caused by: java.lang.IllegalArgumentException: Unrecognized field "x-reprezen-normalization" (class com.reprezen.genflow.api.normal.openapi.RepreZenVendorExtension), not marked as ignorable (6 known properties: "fileUrl", "reason", "typeName", "unresolvableRef", "position", "pointer"])
 at [Source: N/A; line: -1, column: -1] (through reference chain: com.reprezen.genflow.api.normal.openapi.RepreZenVendorExtension["x-reprezen-normalization"])
	at com.fasterxml.jackson.databind.ObjectMapper._convert(ObjectMapper.java:3454)
	at com.fasterxml.jackson.databind.ObjectMapper.convertValue(ObjectMapper.java:3376)
	at com.reprezen.genflow.api.normal.openapi.RepreZenVendorExtension.getFromJsonNode(RepreZenVendorExtension.java:216)
	at com.reprezen.genflow.api.normal.openapi.RepreZenVendorExtension.get(RepreZenVendorExtension.java:127)
	at com.reprezen.genflow.api.normal.openapi.RepreZenVendorExtension.getOrCreate(RepreZenVendorExtension.java:139)
	at com.reprezen.genflow.api.normal.openapi.OpenApiMarkers.markPosition(OpenApiMarkers.java:60)
	at com.reprezen.genflow.api.normal.openapi.OpenApiReferenceProcessor.markPositionsInSection(OpenApiReferenceProcessor.java:395)
	at com.reprezen.genflow.api.normal.openapi.OpenApiReferenceProcessor.markPositions(OpenApiReferenceProcessor.java:381)
	at com.reprezen.genflow.api.normal.openapi.OpenApiReferenceProcessor.inline(OpenApiReferenceProcessor.java:158)
	at com.reprezen.genflow.api.normal.openapi.OpenApiNormalizer.normalizeToJsonInternal(OpenApiNormalizer.java:106)
	at com.reprezen.genflow.api.normal.openapi.OpenApiNormalizer.normalizeToKaizen(OpenApiNormalizer.java:239)
	at com.reprezen.genflow.api.normal.openapi.OpenApiNormalizer.normalizeToKaizen(OpenApiNormalizer.java:264)
	at com.reprezen.genflow.api.openapi3.OpenApi3Source.load(OpenApi3Source.java:52)
	... 9 more
Caused by: com.fasterxml.jackson.databind.exc.UnrecognizedPropertyException: Unrecognized field "x-reprezen-normalization" (class com.reprezen.genflow.api.normal.openapi.RepreZenVendorExtension), not marked as ignorable (6 known properties: "fileUrl", "reason", "typeName", "unresolvableRef", "position", "pointer"])
 at [Source: N/A; line: -1, column: -1] (through reference chain: com.reprezen.genflow.api.normal.openapi.RepreZenVendorExtension["x-reprezen-normalization"])
	at com.fasterxml.jackson.databind.exc.UnrecognizedPropertyException.from(UnrecognizedPropertyException.java:51)
	at com.fasterxml.jackson.databind.DeserializationContext.reportUnknownProperty(DeserializationContext.java:836)
	at com.fasterxml.jackson.databind.deser.std.StdDeserializer.handleUnknownProperty(StdDeserializer.java:1045)
	at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.handleUnknownProperty(BeanDeserializerBase.java:1352)
	at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.handleUnknownVanilla(BeanDeserializerBase.java:1330)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:262)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:125)
	at com.fasterxml.jackson.databind.ObjectMapper._convert(ObjectMapper.java:3449)
	... 21 more
```

**Comment by tedepstein** [comment]
@ghillairet , I am thinking that we need at lease a small suite of tests to cover the integration scenario of live-views, including the normalizer and the three GenTemplates: Documentation, Diagram and Swagger-UI. 

The idea of these tests -- which are really integration tests, not unit tests -- is to simulate the live view processing pipeline. Each test in this suite would start with a document (OAS2 and OAS3, leaving out RAPID-ML for now), and pipe it through the same processes, with the same configurations, as the IDE does. OAS2 and OAS3 documents will go through the normalizer, then live view generation using the appropriate GenTemplate. Once the "rendering" is complete, the tests can check the output to make sure essential content is present as expected.

Am I correct in thinking that this kind of test suite would have caught this error?

**Comment by tedepstein** [comment]
OK, I have just merged `master` into this task branch, and now TC and Travis are both reporting test failures:

```
 com.reprezen.genflow.swagger.doc (9)
 CardinalityTest.testMaxItemsEqualsOne 	
 CardinalityTest.testMaxItemsEqualsZero 	
 CardinalityTest.testMinItemsEqualsOne 	
 CardinalityTest.testMinItemsEqualsZero 	
 CardinalityTest.testMinMaxItems 	
 CardinalityTest.testNoMinItems 	
 PropertyTableTest.testMethodHeader 	
 PropertyTableTest.testMethodPrimtitiveTypes 	
 PropertyTableTest.testResponseReturnType 
```

I think both `Cardinality Test` and `PropertyTableTest` were newly introduced in PR #45, and are probably doing just what I described - testing the output of a two-stage pipeline: `OAS` --> `Normalizer` --> `Doc`.  So maybe my wish came true, retroactively.

**Comment by ghillairet** [comment]
@tedepstein Seems like adding this ROOT ObjectType (https://github.com/RepreZen/GenFlow/pull/46/commits/fc7ca5649958ea0e10a4243d0c22f8b5b98a93ac#diff-78c148ddf3d9aae876450ffc3293ea9dR35) is causing the tests to fail as well as live views. 

It is adding unnecessary reprezen extensions that are causing failures in generators. For example it is adding an extension inside `paths` causing the test doc generator to fail.

```
 "paths": {
    "/pets": {
      "x-reprezen-normalization": {
        "pointer": "/paths/~1pets",
        "position": 0
      },
      "get": {
        "responses": {
          "200": {
            "description": "Ok",
            "schema": {}
            },
            "x-reprezen-normalization": {
              "position": 0
            }
          }
        },
        "x-reprezen-normalization": {
          "pointer": "/paths/~1pets/get",
          "position": 0
        }
      }
    },
// Extension should not be added here
    "x-reprezen-normalization": {
      "position": 2
    }
  }
```

**Comment by tedepstein** [comment]
Leaving the `task/ZEN-4711` branch, in case that makes it easier to revisit the change history for follow-up issue [ZEN-4745](https://modelsolv.atlassian.net/browse/ZEN-4745).

**Pull request merged**